Expose default table expiration time for controlled BQ datasets.

marikomedlock opened the pull request:
Exposed two properties of BigQuery datasets that users do not have permission to modify directly:
- default table expiration time ([ref](https://cloud.google.com/bigquery/docs/updating-datasets#table-expiration))
- default partition expiration time ([ref](https://cloud.google.com/bigquery/docs/updating-datasets#partition-expiration))

The default table expiration time is a request from the solutions team. I included the default partition expiration time also because it was similar and also not modifiable by users directly.

Modified the controlled BQ dataset update endpoint to take an attributes object, similar to the GCS bucket update endpoint. Added two additional properties to the controlled BQ dataset create endpoint.

Wrote tests for:
- (unit) BQ <-> WSM API time conversions
- (unit) step to do the BQ dataset update call
- (connected) create and update flights, with un/defined and invalid expiration times
- (integration) modified the existing BQ dataset lifecycle test to include an instance of the new attributes object

jaycarlton reviewed (approved):
You really followed all the existing conventions well! Thanks for all the tests.

ddietterich reviewed (commented):
Two things to check:
1. Should we use ms instead of secs for lifetime? Is that how other interfaces take the value?
2. IIUC, the case where the dataset that should be there according to WSM, but is not there seems like it should be a 500-class exception.

marikomedlock commented on service/src/main/java/bio/terra/workspace/service/crl/CrlService.java:
```diff
@@ -167,6 +168,35 @@ public boolean canReadBigQueryDataset(
+   */
+  public Dataset getBigQueryDataset(BigQueryCow bigQueryCow, String projectId, String datasetName)
+      throws IOException {
+    return bigQueryCow.datasets().get(projectId, datasetName).execute();
```
I spent a while trying to get that to work, ultimately unsuccessfully. I hit two problems:
- I couldn't mock `execute` because Mockito didn't recognize the expected return value for some reason. It's supposed to return a Dataset, but Mockito wanted me to return an Optional. Not sure why -- maybe an inheritance problem?
- I couldn't mock a `Dataset` because it's a final class. There is an experimental feature in Mockito to enable this, but unclear if that will be supported long-term.

marikomedlock commented on service/src/main/java/bio/terra/workspace/service/resource/controlled/BqApiConversions.java:
```diff
@@ -0,0 +1,57 @@
+    // since we're trying to preserve the exiting state in case we need to put it back during an
+    // undo, convert an existing null expiration time to a 0, so that it gets removed again
+    if (updateParams.getDefaultTableLifetime() == null) {
+      updateParams.setDefaultTableLifetime(0);
```
Agreed. I chose the special zero to mimic what the `bq` CLI does for these properties.

marikomedlock commented on service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/update/UpdateBigQueryDatasetStep.java:
```diff
@@ -0,0 +1,118 @@
+      // get the existing dataset
+      Dataset existingDataset = crlService.getBigQueryDataset(bigQueryCow, projectId, datasetId);
+      if (existingDataset == null) {
+        logger.info("No dataset found to update with id {}. Treating as success.", datasetId);
```
Yeah I thought this was weird, too. This is what the `UpdateGcsBucketStep` does ([code ref](https://github.com/DataBiosphere/terra-workspace-manager/blob/034eac988cffa410b9b675227a4fded6abfd3968/service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/update/UpdateGcsBucketStep.java#L78)), so I copied the same behavior.

Would you like me to change both places to return step failure instead?

ddietterich replied:
Yes, I think so.

marikomedlock replied:
Done.

marikomedlock commented on service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/update/RetrieveBigQueryDatasetCloudAttributesStep.java:
```diff
@@ -0,0 +1,72 @@
+      existingDataset = bigQueryCow.datasets().get(projectId, datasetId).execute();
+      if (existingDataset == null) {
+        logger.info("No dataset found with id {}.", datasetId);
+        return new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL, null);
```
Makes sense. Changed to pass an `IllegalStateException` to the step result.

marikomedlock commented on service/src/main/java/bio/terra/workspace/service/resource/controlled/BqApiConversions.java:
```diff
@@ -0,0 +1,57 @@
+import javax.annotation.Nullable;
+
+/** Utility methods for converting BQ dataset objects between WSM formats and GCP formats. */
+public class BqApiConversions {
```
Renamed to `BigQueryApiConversions` and the associated test class.

marikomedlock commented on service/src/main/java/bio/terra/workspace/service/resource/controlled/BqApiConversions.java:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  /**
+   * Helper method to convert ms -> sec, because the GCP API uses ms and the WSM API uses sec. The
```
GCP isn't really consistent either: cloud console uses days, `bq` CLI uses seconds, API uses milliseconds. The solutions team use case is days (they just want an auto-delete flag so test outputs get cleaned up), but I picked seconds to be consistent with the `bq` CLI. I don't have a strong opinion on this though, so happy to change to any of these units.

ddietterich replied:
I would vote for days in WSM REST API, just because it is consistent with the lifetime spec for buckets.
But it is also fine the way it is.

marikomedlock commented on service/src/main/java/bio/terra/workspace/service/resource/controlled/BqApiConversions.java:
```diff
@@ -0,0 +1,57 @@
+   * bq command line also uses sec.
+   */
+  public static Integer fromBqExpirationTime(@Nullable Long ms) {
+    return ms == null ? null : Math.toIntExact(ms / 1000);
```
😄

marikomedlock commented:
@jaycarlton @ddietterich changes round two:
- updated the places where a dataset/bucket is not found to throw an `IllegalStateException` instead of failing silently or without an exception.
- added handling for a 400 error from GCP when the expiration times are invalid (<3600 sec for table, <0 sec for partition), and a test for this.
thanks!

ddietterich reviewed (approved):
LGTM - thank you!

marikomedlock commented on service/src/main/java/bio/terra/workspace/service/resource/controlled/ControlledResourceService.java:
```diff
@@ -191,6 +193,31 @@ public ControlledBigQueryDatasetResource createBigQueryDataset(
+      @Nullable String resourceDescription) {
+    final String jobDescription =
+        String.format(
+            "Update controlled BigQuery Dataset resource %s; id %s; name %s",
```
agreed, fixed.
